platform/pulp/platform/tasks/task_system.py

```diff
@@ -1,4 +1,548 @@
+    find_worker function.
+
+    :param name:          The name of the task to be called
+    :type name:           basestring
+    :param inner_task_id: The UUID to be set on the task being called. By providing
+                          the UUID, the caller can have an asynchronous reference to the inner task
+                          that will be dispatched.
+    :type inner_task_id:  basestring
+    :param resource_id:   The name of the resource you wish to reserve for your task. The system
+                          will ensure that no other tasks that want that same reservation will run
+                          concurrently with yours.
+    :type  resource_id:   basestring
+
+    :return: None
+    """
+    while True:
+        # Find a worker who already has this reservation, it is safe to send this work to them
+        try:
+            worker = ReservedResource.objects.get(resource=resource_id).worker
+        except ReservedResource.DoesNotExist:
+            pass
+        else:
+            break
+
+        try:
+            worker = _get_unreserved_worker()
+        except Worker.DoesNotExist:
+            pass
+        else:
+            break
+
+        # No worker is ready for this work, so we need to wait
+        time.sleep(0.25)
+
+    task_status = TaskStatus.objects.get(id=task_id)
+    ReservedResource(task=task_status, worker=worker, resource=resource_id).save()
+
+    inner_kwargs['routing_key'] = worker.name
+    inner_kwargs['exchange'] = DEDICATED_QUEUE_EXCHANGE
+    inner_kwargs['task_id'] = task_id
+
+    try:
+        celery.tasks[name].apply_async(*inner_args, **inner_kwargs)
+    finally:
+        _release_resource.apply_async((task_id, ), routing_key=worker.name,
+                                      exchange=DEDICATED_QUEUE_EXCHANGE)
+
+
+def _is_worker(worker_name):
+    """
+    Strip out workers that should never be assigned work. We need to check
+    via "startswith()" since we do not know which host the worker is running on.
+    """
+
+    if worker_name.startswith(SCHEDULER_WORKER_NAME) or \
+       worker_name.startswith(RESOURCE_MANAGER_QUEUE):
+        return False
+    return True
+
+
+def _get_unreserved_worker():
+    """
+    Find an unreserved Worker
+
+    Return the Worker instance that has no ReservedResource associated with it. If all workers have
+    ReservedResource relationships a Worker.DoesNotExist pulp.platform.model.Worker exception is
+    raised.
+
+    This function also provides randomization for worker selection.
+
+    :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with them.
+
+    :returns:          The Worker instance that has no reserved_resource
+                       entries associated with it.
+    :rtype:            pulp.platform.model.Worker
+    """
+    free_workers = [Worker.objects.annotate(Count('reservations')).filter(reservations__count=0)]
+    if len(free_workers) == 0:
+        raise ReservedResource.DoesNotExist()
```

Comment:
I did mean `Worker.DoesNotExist` thanks for point this out. It will be updated on my next push.